File: movedrone_setup_assistant/src/movedrone_setup_assistant/parameter_getters/vector3d.py
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from typing import List, Tuple

from dh_rqt_tools.widgets import DoubleSpinBox

from .base import ParamGetterWidget
from ..const import *


class ParamGetterWidget_Vector3d(ParamGetterWidget):

    value_changed = pyqtSignal(float, float, float)

    def __init__(
        self,
        param_name: str,
        description_text: str = None,
        minimum: List[float] = [-1e+9] * 3,
        maximum: List[float] = [+1e+9] * 3,
        single_step: List[float] = [1.] * 3,
        default: List[float] = [0.] * 3,
        suffix: str = "",
    ) -> None:
        assert len(minimum) == 3
        assert len(maximum) == 3
        assert len(single_step) == 3
        assert len(default) == 3

        super().__init__(param_name, description_text)

        self._cols = QHBoxLayout()
        self._rows.addLayout(self._cols)

        self._x = DoubleGetter("x", minimum[0], maximum[0], single_step[0], default[0], suffix)
        self._cols.addWidget(self._x)

        self._y = DoubleGetter("y", minimum[1], maximum[1], single_step[1], default[1], suffix)
        self._cols.addWidget(self._y)

        self._z = DoubleGetter("z", minimum[2], maximum[2], single_step[2], default[2], suffix)
        self._cols.addWidget(self._z)

        self._x.data.valueChanged.connect(self._on_value_changed)
        self._y.data.valueChanged.connect(self._on_value_changed)
        self._z.data.valueChanged.connect(self._on_value_changed)

    def get(self) -> Tuple[float, float, float]:
        return self._x.get(), self._y.get(), self._z.get()

    def set(self, x: float, y: float, z: float) -> None:
        self._x.data.setValue(x)
        self._y.data.setValue(y)
        self._z.data.setValue(z)

    @pyqtSlot(float)
    def _on_value_changed(self, value: float) -> None:
        self.value_changed.emit(self._x.get(), self._y.get(), self._z.get())


class DoubleGetter(QWidget):

    def __init__(
        self,
        name: str,
        minimum: float,
        maximum: float,
        single_step: float,
        default: float,
        suffix: str,
    ) -> None:
        assert minimum < maximum
        assert single_step > 0.

        super().__init__()

        self._cols = QHBoxLayout()
        self.setLayout(self._cols)

        label = QLabel(name + ":")
        label.setFont(QFont("Default", pointSize=BODY_PSIZE))
        label.setAlignment(Qt.AlignRight)
        self._cols.addWidget(label)

        self.data = DoubleSpinBox()
        self.data.setMinimum(minimum)
        self.data.setMaximum(maximum)
        self.data.setSingleStep(single_step)
        if default is not None:
            assert minimum <= default <= maximum
            self.data.setValue(default)
        self.data.setSuffix(suffix)
        self._cols.addWidget(self.data)

    def get(self) -> float:
        return self.data.value()

File: tobas_gui/tobas_setup_assistant/src/param_getters/vector3d.cpp
#include "tobas_setup_assistant/param_getters/vector3d.hpp"

namespace gui
{
namespace sa
{
ParamGetterWidget_Vector3d::ParamGetterWidget_Vector3d(const QString& param_name, const QString& description_text)
  : super(param_name, description_text)
{
  const auto cols = new QHBoxLayout();
  rows_->addLayout(cols);

  x_ = new DoubleGetter("x");
  cols->addWidget(x_);

  y_ = new DoubleGetter("y");
  cols->addWidget(y_);

  z_ = new DoubleGetter("z");
  cols->addWidget(z_);

  connect(x_, QOverload<double>::of(&DoubleGetter::valueChanged), this, &self::onValueChanged);
  connect(y_, QOverload<double>::of(&DoubleGetter::valueChanged), this, &self::onValueChanged);
  connect(z_, QOverload<double>::of(&DoubleGetter::valueChanged), this, &self::onValueChanged);
}

Eigen::Vector3d ParamGetterWidget_Vector3d::getValue() const
{
  return { x(), y(), z() };
}

bool ParamGetterWidget_Vector3d::setValue(const Eigen::Vector3d& src)
{
  x_->setValue(src.x());
  y_->setValue(src.y());
  z_->setValue(src.z());
  return true;
}

void ParamGetterWidget_Vector3d::setDecimals(int decimals)
{
  x_->setDecimals(decimals);
  y_->setDecimals(decimals);
  z_->setDecimals(decimals);
}

void ParamGetterWidget_Vector3d::setMinimum(double minimum)
{
  x_->setMinimum(minimum);
  y_->setMinimum(minimum);
  z_->setMinimum(minimum);
}

void ParamGetterWidget_Vector3d::setMaximum(double maximum)
{
  x_->setMaximum(maximum);
  y_->setMaximum(maximum);
  z_->setMaximum(maximum);
}

void ParamGetterWidget_Vector3d::setSingleStep(double single_step)
{
  x_->setSingleStep(single_step);
  y_->setSingleStep(single_step);
  z_->setSingleStep(single_step);
}

void ParamGetterWidget_Vector3d::setSuffix(const QString& suffix)
{
  x_->setSuffix(suffix);
  y_->setSuffix(suffix);
  z_->setSuffix(suffix);
}

double ParamGetterWidget_Vector3d::x() const
{
  return x_->getValue();
}

double ParamGetterWidget_Vector3d::y() const
{
  return y_->getValue();
}

double ParamGetterWidget_Vector3d::z() const
{
  return z_->getValue();
}

void ParamGetterWidget_Vector3d::onValueChanged(double)
{
  Q_EMIT valueChanged({ x(), y(), z() });
}
}  // namespace sa
}  // namespace gui
